newton/examples/mpm/example_mpm_twoway_coupling.py

```diff
@@ -0,0 +1,404 @@
+        # z height to drop shapes from
+        drop_z = 2.0
+
+        # layout: spawn shapes near the same XY so they collide/stack
+        offsets_xy = [
+            (0.00, 0.00),
+            (0.10, 0.00),
+            (-0.10, 0.00),
+            (0.00, 0.10),
+            (0.00, -0.10),
+            (0.10, 0.10),
+            (-0.10, 0.10),
+            (0.10, -0.10),
+            (-0.10, -0.10),
+            (0.15, 0.00),
+            (-0.15, 0.00),
+            (0.00, 0.15),
+        ]
+        offset_index = 0
+        z_index = 0
+        z_separation = 0.6  # vertical spacing to avoid initial overlap
+
+        # generate a few boxes with varying sizes
+        # boxes = [(0.45, 0.35, 0.25)]  # (hx, hy, hz)
+        boxes = [
+            (0.25, 0.35, 0.25),
+            (0.25, 0.25, 0.25),
+            (0.3, 0.2, 0.2),
+            (0.25, 0.35, 0.25),
+            (0.25, 0.25, 0.25),
+            (0.3, 0.2, 0.2),
+        ]  # (hx, hy, hz)
+        for box in boxes:
+            (hx, hy, hz) = box
+
+            ox, oy = offsets_xy[offset_index % len(offsets_xy)]
+            offset_index += 1
+            pz = drop_z + float(z_index) * z_separation
+            z_index += 1
+            body = builder.add_body(
+                xform=wp.transform(p=wp.vec3(float(ox), float(oy), pz), q=wp.normalize(wp.quatf(0.0, 0.0, 0.0, 1.0))),
+                mass=75.0,
+            )
+            builder.add_shape_box(body, hx=float(hx), hy=float(hy), hz=float(hz))
+
+    def _emit_particles(self, sand_builder: newton.ModelBuilder, voxel_size: float):
+        # ------------------------------------------
+        # Add sand bed (2m x 2m x 0.5m) above ground
+        # ------------------------------------------
+
+        particles_per_cell = 3.0
+        density = 2500.0
+
+        bed_lo = np.array([-1.0, -1.0, 0.0])
+        bed_hi = np.array([1.0, 1.0, 0.5])
+        bed_res = np.array(np.ceil(particles_per_cell * (bed_hi - bed_lo) / voxel_size), dtype=int)
+
+        # spawn particles on a jittered grid
+        Nx, Ny, Nz = bed_res
+        px = np.linspace(bed_lo[0], bed_hi[0], Nx + 1)
+        py = np.linspace(bed_lo[1], bed_hi[1], Ny + 1)
+        pz = np.linspace(bed_lo[2], bed_hi[2], Nz + 1)
+        points = np.stack(np.meshgrid(px, py, pz)).reshape(3, -1).T
+
+        cell_size = (bed_hi - bed_lo) / bed_res
+        cell_volume = np.prod(cell_size)
+        radius = float(np.max(cell_size) * 0.5)
+        mass = float(np.prod(cell_volume) * density)
+
+        rng = np.random.default_rng()
+        points += 2.0 * radius * (rng.random(points.shape) - 0.5)
+        vel = np.zeros_like(points)
+
+        sand_builder.particle_q = points
+        sand_builder.particle_qd = vel
+        sand_builder.particle_mass = np.full(points.shape[0], mass)
+        sand_builder.particle_radius = np.full(points.shape[0], radius)
+        sand_builder.particle_flags = np.ones(points.shape[0], dtype=int)
```

Comment:
Can this code be used to replace what we have now in `ModelBuilder.add_particle_grid()`?
This implementation is certainly a lot faster than those for loops.